src/auxinfo/proof.rs

```diff
@@ -142,3 +183,115 @@ impl AuxInfoProof {
         Ok(())
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{paillier::prime_gen, protocol::SharedContext, utils::testing::init_testing};
+    use rand::Rng;
+
+    fn random_auxinfo_proof<R: RngCore + CryptoRng>(
+        rng: &mut R,
+        test_code: impl FnOnce(CommonInput, AuxInfoProof) -> Result<()>,
+    ) -> Result<()> {
+        let sid = Identifier::random(rng);
+        let rho = rng.gen();
+        let setup_params = VerifiedRingPedersen::gen(rng, &()).unwrap();
+        let (p, q) = prime_gen::get_prime_pair_from_pool_insecure(rng).unwrap();
+        let modulus = &p * &q;
+        let shared_context = SharedContext::random(rng);
+        let common_input = CommonInput::new(&shared_context, sid, rho, &setup_params, &modulus);
+        let proof =
+            AuxInfoProof::prove(rng, &common_input.clone(), &p.clone(), &q.clone()).unwrap();
+        test_code(common_input, proof)
+    }
+
+    #[test]
+    fn auxinfo_proof_verifies() -> Result<()> {
+        let mut rng = init_testing();
+        let sid = Identifier::random(&mut rng);
+        let rho = rng.gen();
+        let setup_params = VerifiedRingPedersen::gen(&mut rng, &())?;
+        let (p, q) = prime_gen::get_prime_pair_from_pool_insecure(&mut rng).unwrap();
+        let modulus = &p * &q;
+        let shared_context = SharedContext::random(&mut rng);
+        let common_input = CommonInput::new(&shared_context, sid, rho, &setup_params, &modulus);
+        let proof = AuxInfoProof::prove(&mut rng, &common_input, &p, &q)?;
+        assert!(proof.verify(&common_input).is_ok());
+        Ok(())
+    }
+
+    #[test]
+    fn each_constituent_proof_must_be_valid() -> Result<()> {
+        let mut rng = init_testing();
+        let mut rng_clone = rng.clone();
```

Comment:
Here's the issue! When you clone an RNG you also clone its internal state. So instead of getting two different-stated RNGs that produce two different proofs, we're using the same RNG to make `proof` and `proof1`, so they're the same and mixing their fields still produces a valid proof.

Instead, make a new RNG by seeding it with the existing one. That way it'll be linked to the one we made with `init_testing`, so we maintain reproducibility, but will have a different state.
```rust
let mut rng2 = StdRng::from_rng(&mut rng).unwrap();
 ```